fw/src/mppt/global/simulated_annealing.hpp

```diff
@@ -11,69 +11,99 @@
+        if (temperature > 0.1) {
+            float current_power = array_voltage * array_current;
+            float best_power = current_power;
+            float best_voltage = array_voltage;
 
-            if (new_power > max_power) {
-                max_power = new_power;
-                reference_voltage = new_voltage;
-            } else {
-                float delta_power = new_power - max_power;
-                if (exp(delta_power / current_temperature) > distribution(engine)) {
-                    reference_voltage = new_voltage;
+            for (int i = 0; i < NUM_ITERATIONS; i++) {
+                float new_voltage = array_voltage + get_random_value(-1.0, 1.0);
+                new_voltage = std::max(0.0f, std::min(new_voltage, MAX_INP_VOLT));
+                float new_power = new_voltage * array_current;
+
+                if (new_power > best_power) {
+                    best_power = new_power;
+                    best_voltage = new_voltage;
+                } else {
+                    float acceptance_probability = std::exp((new_power - current_power) / temperature);
+                    if (get_random_value(0.0, 1.0) < acceptance_probability) {
+                        best_power = new_power;
+                        best_voltage = new_voltage;
+                    }
                 }
             }
 
-            current_temperature *= cooling_rate;
+            reference_voltage = best_voltage;
+            temperature *= COOLING_RATE;
+        } else {
+            reference_voltage = pandO->step_algorithm();
         }
+
+        prev_array_power = array_voltage * array_current;
+        prev_array_voltage = array_voltage;
     }
 
-    void reset_state(void) override {
-        MPPT::reset_state();
-        current_temperature = initial_temperature;
-        max_power = 0.0;
+    void reset_state(void) {
+        reference_voltage = 0.0;
+        prev_array_voltage = 0.0;
+        prev_array_power = 0.0;
+        temperature = INITIAL_TEMP;
+        pandO->reset_state();
     }
 
 protected:
     float array_voltage;
     float array_current;
-    float max_power = 0.0;
+    float battery_voltage;
+    float battery_current;
 
-    float initial_temperature = 25.0;
-    float final_temperature = 0.4;
-    float cooling_rate = 0.75;
-    float current_temperature;
-
-    std::mt19937 engine;
-    std::uniform_real_distribution<float> distribution;
+    float prev_array_voltage;
+    float prev_array_power;
+    float temperature;
+    PandO* pandO;
+    std::mt19937 gen;
 
 private:
-    float generate_new_voltage(float current_voltage) {
-        float voltage_variation = distribution(engine);
-        return current_voltage + voltage_variation;
-    }
-
-    float calculate_power(float voltage, float current) {
-        return voltage * current;
+    float get_random_value(float min, float max) {
+        std::uniform_real_distribution<float> dist(min, max);
+        return dist(gen);
```

Comment:
Can we remove the dependency on STD?
MBED should support cryptographic generation that uses some aspect of the device, i think we can rework this to use that:

https://forums.mbed.com/t/how-to-generate-random-numbers-t-rng/7393/2

And if all else fails, read from the internal temperature adc and use that.
I'm not convinced that for this algorithm to be useful that it needs to be truly random.